Throttlebased EzLanding

tbolin opened the pull request:
An alternative, throttle based, method for reducing "bounciness". 
The original method as described by @ctzsnooze in #11998 uses motor output clipping.  
This PR also contains that method for easier comparison. 

The method implemented here tries to solve some issues with the motor clipping method:

* Further reduced bounciness when landing at just below hover throttle
* iTerm windup should be handled better
* Cross talk should be reduced since the outputs are scaled, rather than clipped
* Better interaction with dynamic idle, possibly preventing some of the instability on throttle chops

## Some definitions
* **effective throttle** the average motor output as a percentage of maximum motor output.
* **requested throttle** the throttle requested by the pilot.
* **PID motor outputs** the motor output values requested by the PID controllers.

## How the method works compared to airmode off and the previous PR
Bounciness during landings basically happens since large PID outputs from collisions causes the throttle to be adjusted upwards.
In the original PR the effective throttle is prevented from increasing beyond a limit by simply clipping the motor outputs if the throttle is low and the other sticks are close to centered.
The method in this PR instead limits how much the throttle can be adjusted upwards.

Unlike the old (pre 4.4) airmode off this method will scale the PID motor outputs, instead of clipping them if they do not fit.
Unlike the new 4.4 airmode off the PID motor outputs will only be scaled if they would otherwise be clipped.

**Algorithm outline** 
1. Calculate max allowable throttle increase according to some formula.
2. Calculate the "airmode throttle" the same way that it is currently done. 
3. If the airmode throttle is higher than the allowable throttle limit, set the throttle to that limit and scale the motor output from the PIDs so that the lowest value + the calculated throttle = minimum allowed motor output.
If the airmode throttle is below the limit, don't scale anything.
4. Make sure that the anti windup is aware of the reduced authority range.

*some formula* here is almost the same formula that is used in the original PR to calculate the limit motor output limit.
The one change is that throttle is not factored in directly.
Instead the maximum of the requested throttle and the output from the formula is used.
Since the formula have a minimum output at `ez_landing_limit`, the PIDs will have a bit of wiggle room even with requested throttle at 0 and stick centered. 

## Usage 
To activate the algorithm `mixer_type` should be set to `EZLANDING`

For testing: set `debug_mode` to `EZLANDING`

### Settings
* `ez_landing_limit`  _Default: 5, Range: 0-75_. the mixer will always be allowed to increase the throttle to at least this value in percent.
Increase if the quad feels unstable at zero throttle, decrease if the quad feels too bouncy with the RPY sticks centered. 
* `ez_landing_threshold` _Default: 25, Range: 0-200_.  Deflecting the roll, pitch or yaw stick this much in percent will effectively give the mixer full authority over the throttle (i.e. as whit airmode on with the Legacy mixer type). The authority of the mixer scales linearly when the max deflection is bellow this value. 

### Debug values
* **Field 0** Scaling of the motor range. If this is 10000 the throttle and motor output range is currently the same as it would be with the Legacy mixer and airmode on. If  the value is lower the motor output range is being scaled down to avoid increasing the throtte.
* **Field 1** Same as field 1 but if the throttle set point was 0.
* **Field 2** The maximum throttle that the mixer is allowed to request.
* **Field 3** Effective throttle, the throttle that the mixer is actually requesting.

### Old Usage
To activate the algorithm `mixer_type` should be set to `EZLANDING_THROTTLE`
For comparison the algorithm from the previous PR can be activated by setting `mixer_type` to `EZLANDING_CLIP`.

The default values for `ez_landing_limit` and `ez_landing_threshold` are the same as in the original PR (0.1 and 0.2).
The throttle based algorithm uses the values slightly differently so if `EZLANDING_THROTTLE` is set, they will be automatically adjusted to make the authority available to the PIDs as close as possible to the original PR (ez_landing_limit will be halved and ez_landing_threshold doubled when used in the code).

Setting `mixer_type` to `LEGACY` should turn of both modes and make the quad behave as normal.

## Testing instructions
See the original PR for basic testing instructions.

Pay special attention to: 
* instability at throttle chops and long dead drops
* Bounciness when landing at slightly below hover throttle

## Known issues
* Most likely not compatible with 3D quads in the current shape
* Sudden onset of PID authority. Set-point or set-point movement might be a better way to estimate how much priority the PIDs should be given 

Please mark this PR with *Testing Required* and *Don't Merge*

blckmn commented:
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> FAIL
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> FAIL
 - at least one `RN:` label found -> FAIL
 - `Tested` label found -> FAIL
 - assigned to an approver -> FAIL
 - approver count at least three -> FAIL

tbolin commented:
There was an issue with the anti windup not activating until after the point of saturation. Should hopefully be fixed now.

haslinghuis commented on src/main/fc/rc.c:
```diff
@@ -132,6 +133,11 @@ float getRcDeflectionAbs(int axis)
     return rcDeflectionAbs[axis];
 }
 
+float getMaxRcDeflectionAbs(void)
+{
```
```suggestion
float getMaxRcDeflectionAbs(void) {
```

https://betaflight.com/docs/development/CodingStyle#opening-braces

haslinghuis commented on src/main/fc/rc.c:
```diff
@@ -132,6 +133,11 @@ float getRcDeflectionAbs(int axis)
     return rcDeflectionAbs[axis];
 }
 
+float getMaxRcDeflectionAbs(void)
+{
```
```suggestion
float getMaxRcDeflectionAbs(void) {
```

Please see: https://betaflight.com/docs/development/CodingStyle#opening-braces

tbolin replied:
Functions should have the opening brace on the following line according to the coding style docs, so the current version looks correct to me.

haslinghuis commented on src/main/flight/mixer.c:
```diff
@@ -619,10 +682,23 @@ FAST_CODE_NOINLINE void mixTable(timeUs_t currentTimeUs)
     motorMixRange = motorMixMax - motorMixMin;
-    if (mixerConfig()->mixer_type > MIXER_LEGACY) {
+
+    switch (mixerConfig()->mixer_type)
+    {
```
```suggestion
    switch (mixerConfig()->mixer_type) {
```

tbolin replied:
should be fixed now

haslinghuis reviewed (changes requested):
Some minor code styling issues

tbolin commented:
> Some minor code styling issues

Thank! I'll fix those soon.

A quick question about the new auto-squich on merge: what does the commit message of the squished commit say? 
I don't want to end up with a commit that says something useless like "Fixed minor style issues" for the whole thing.

tbolin commented:
I'l repeat my recent post in the testers forum on discord here:
 
Currently I'm looking at improving the behavior in angle mode at min throttle. 

Since the ezlanding strength is based on stick deflection and not the actual rate set point used by the rate PID loop the ezlanding strength is likely to be inappropriate for the requested movement.
I've been able to show a bit of down scaling at zero throttle when wobbling back and forth.

The most obvious solution is to use the rate pid setpoint, however currently it's easier said than done, and exposing it might interfere with work on the angle mode improvements.

Another thing I'm thinking about is to also use feed forward to calculate how much authority the PIDs are given. A similar approach with throttle might also be useful for preventing wobbling at throttle cuts as reported by Sek101, without having to increase the baseline authority. 
Currently I'm leaning towards saving these for a later PR.

haslinghuis commented:
All commit messages are preserved as we squash commits automatically on merge.

tbolin commented:
Another repeat from the discord so it's searchable:
I realized that rate PID setpoint won't work for angle mode. It's based on sensor readings so if the quad receives a nudge the limit will be raised even if the sticks are at neutral.

I think it's best to wait until the angle mode changes in PR #12231  are merged into main and see if the behavior in angle mode needs to be improved after that.

tbolin commented:
Now with the angle mode PR merged (and a whoop with a flash chip) I was finally able to do a comparison of the throttle based and clipping based EzLanding versions.

[ezland_sir_quack_clip_angle.zip](https://github.com/betaflight/betaflight/files/11006308/ezland_sir_quack_clip_angle.zip)
[ezland_sir_quack_throttle_angle.zip](https://github.com/betaflight/betaflight/files/11006309/ezland_sir_quack_throttle_angle.zip)

I'll get back with an analysis of the logs soon.

tbolin commented:
"Soon" is apparently two months.

### Dynamic idle (and thrust linear)

At startup with dynamic idle limited to 5%, and with throttle SP at 0. 
The clipping based mode is locked at 10.56% (should really be 10, not sure where the extra 0.56% come from).
Clip does not adjust for dynamic idle (or any static idle value either).

The throttle based mode allows at least 20.56% maximum in the log.
Throttle should limit the average (not the max) of the motors to idle value + 5% with throttle at 0 and no stick input.
Here that would mean 10%, while the maximum average in the log is about 12.5%.
The extra percent are probably due to an interaction with thrust_linear (set to 60 in both logs).
The effect from thrust linear in the throttle based mode is probably a good thing since quads which use a lot of thrust linear have less authority at low rpm.
So being affected by thrust linear will allow a default ez_land_limit to be relevant to more quads.

### Zero throttle stability
The clipping based mode saturates and starts drifting in some dead drops.
Since dynamic idle is not accounted for the available authority will slowly decrease during the drop. 
During saturation the I-term also keeps building, and the motor outputs look quite wobbly when the throttle is raised again, compared to without saturation. 

Clipping the motor output vector might also lead introduction of errors on other axes.

The throttle based mode did not saturate during any dead drops, since the `ez_land_limit` is added on top of the motor output from dynamic idle.
Not a lot more motor output was used, and I can't find any significant drift in these logs.

### General loss of authority
I can only find one instance in the log where the throttle based mode scales back authority where it shouldn't.
At 27.058 in part two of the log the stick snaps back from just under 50% roll to 0% in about 25 ms, causing the motor output difference to be scaled down for about 70 ms. 
The only effect on the quad appears to be that it leveled out a bit slower than desired.
The I-term remained constant while the output difference was scaled down, so the anti windup appears to be working as expected now.

I think this could be improved limiting how slow the maximum allowed authority is allowed to decrease, but from what I have seen it's such a minor issue that it can be added in a later PR if needed. 

### Bounciness
Subjectively both modes work well for reducing bounciness (to the degree that it feels weird when I fly without them). 
In both tests I accidentally hit the ceiling, causing the quad to flip, and then straighten up when I increased throttle again.
I barely registered what happened and kept flying.

The throttle based mode should have a slight advantage with the better windup handling and better mapping between the desired throttle and the allowed throttle (e.g. if the quad hovers at 20% the throttle based mode will not allow the throttle to rise above that with RPY sticks centered, while the clipping based mode likely will).
I think I can feel a slight difference with the throttle based mode bouncing less when coming in for soft landings, but I wouldn't bet on that I could tell them apart in a blind test.

## Suggested way forward 
Based on this testing I'm going to remove the clipping based mode, and keep the throttle mode. 
I also think that the name of the feature should be changed to *anti_bounce,* since it makes the quad less bouncy, not just easier to land.

ctzsnooze commented:
I really apologise for not trying out or reviewing this PR.  I fully intend to once we have GPS Rescue issues resolved.

tbolin commented:
I'm going to go ahead with the removal of the clipping based mode and renaming soon.

tbolin commented:
Updated the code with some final refactoring and a slight change to the settings to make the ez_landing_threshold a bit more intuitive.
Also updated the OP with a description of the settings and descriptions of the debug fields.

tbolin commented:
Mostly bumping so the PR doesn't get marked as inactive and let people know it's not abandoned.
As far as I'm concerned this PR is ready to merge functionality wise.
The remaining question is if the feature should be renamed or not.

tbolin commented:
Just preserving some testing results from discord here, posted on 12 June 2023
>Machinehead: @ctzsnooze PR #12094 testing results. After 3 test flights and a series of landing tests each flight, I concur that this PR reduces bounciness greatly. The quad immediately settles down to idle speed after a few very small bounces on a harder landing in acro mode as well as angle mode. The quad settles right down to idle speed on a nice soft more controlled landing with no noticeable reaction to ground contact in either acro or angle mode as well. The test quad has a very high thrust to weight radio and this used to be an issue on this quad. It's target is the MATEKF405STD_CLONE. I always had to disarm before touching the ground or it would bounce up many feet, sometimes in a bad direction. Causing the craft to be a danger. It also flew very well during testing, just as good as any other firmware that is current.

KarateBrot commented on src/main/flight/mixer.h:
```diff
@@ -71,6 +71,8 @@ typedef enum mixerType
     MIXER_LEGACY = 0,
     MIXER_LINEAR = 1,
     MIXER_DYNAMIC = 2,
+    MIXER_EZLANDING_THROTTLE = 3,
+    MIXER_EZLANDING_CLIP = 4,
```
```suggestion
    MIXER_EZLANDING = 3,
```

tbolin replied:
fixed

KarateBrot commented on src/main/flight/mixer.c:
```diff
@@ -668,10 +723,21 @@ FAST_CODE_NOINLINE void mixTable(timeUs_t currentTimeUs)
         applyMixerAdjustmentLinear(motorMix, airmodeEnabled);
-    } else {
+        break;
+    case MIXER_EZLANDING_THROTTLE:
```
```suggestion
    case MIXER_EZLANDING:
```

tbolin replied:
..and fixed. Thanks!

KarateBrot reviewed (commented):
I guess this was left over from old code

ctzsnooze commented:
@tbolin would it be OK to hold this PR over to 4.6?  It can still be flown, but I will now be very busy getting 4.5 ready.  I can give proper attention to the PR once 4.5 is released.

tbolin commented:
Well, It's not like I really have a say in the matter.

ctzsnooze commented:
Well I guess if other people were actively providing feedback and supporting its inclusion that would make a difference.

haslinghuis commented:
Really need a rebase here

ctzsnooze commented:
Thanks @tobolin that was my exact concern.
The latching of `airmodeIsActivated` should occur from the time when the throttle goes above the airmode activate threshold, and remain in place until disarm, or at least that's how I thought it was supposed to work, even at zero throttle.
However the code didn't read quite like that, when working on #13127 the changes there retained iTerm at zero throttle.
I would appreciate your advice and will test this further.
Thanks for getting back to me.

tbolin commented:
@ctzsnooze I checked and, yes the flag still seems to latch. Both the i-term reset and dynamic idle behaves exactly as expected with the flag latching. Good but confusing. I'll have to take another look at that code some day, but for now everything seems to be working as intended.

ctzsnooze commented:
@tbolin - hmm... no, master appears to lose iTerm at zero throttle, with or without this PR.  It was OK on 28/11.

haslinghuis commented:
@tbolin the iterm bug should be resolved. Please rebase and test again.

tbolin commented:
@haslinghuis rebased and tested (with a local build this time). The latching seems to work as expected. 

[dynidle_and_iterm_with_ezland_on.zip](https://github.com/betaflight/betaflight/files/13604584/dynidle_and_iterm_with_ezland_on.zip)

ctzsnooze commented:
I've built and flashed this code to a test setup.

I noticed that iTerm is zero when the pitch/roll stick is centered, and throttle is low.  This will definitely help prevent bouncy landings where iTerm can be a contributor, and help 'un-sticking' a whoop from the wall... but it will make 'perching' very difficult, since adequate iTerm at low throttle is essential to hold the quad at the requested angle.

I also found something very strange that I couldn't really explain. Normally, under all zero throttle conditions, when I roll right, motors 3 and 4 increase.  When I then hold a while, and sweep left across centre, the driven motors shift across until 1 and 2 are higher, when the roll stick is to the left of centre.

But if this PR's version of EZLanding is flashed, something very strange happens once airmode is active. If I activate airmode, put throttle to zero, and move the stick out to the right side, motors 3 and 4 spin up like normal.  However, if I then try to centre the stick, motors 3 and 4 get *stronger and stronger* !!  Very much NOT what I'd want.  Then only once I get the stick back across centre do they finally slow down, and go the other way.

This makes for completely the opposite effect from what is intended, I would have thought.  I don't know if this is a practical issue, but if I was correcting my position with sticks to the right, I would not want the motors to go harder when I was going back to centre, but that's what happens.

Here are two videos, zipped, trying to explain what I found.  Without seeing the stick inputs it may not be sensible, but is easily replicated on any quad with this PR and props off.

[NormalMixer.mov.zip](https://github.com/betaflight/betaflight/files/13608708/NormalMixer.mov.zip)
[ExLandingMixer.mov.zip](https://github.com/betaflight/betaflight/files/13608719/ExLandingMixer.mov.zip)

ctzsnooze commented:
My motor_output_limit method retained iTerm.  Hence it didn't entirely stop the bounce on landing, especially if the quad landed at an angle, because iTerm wasn't gone.  Since the motors couldn't go very high, the bad iTerm effects were markedly reduced.  There was enough iTerm to perch in a stable manner, and hold attitude on the ground if it was a battery-below quad, but not enough to go totally nuts.  Also, moving the sticks always gave consistent results, because iTerm was retained like normal.  It was impossible to flyaway because motor_output was very limited when sticks were centered and throttle zero.

I'm not sure why this PR gets that strange behaviour once airmode is active with small stick inputs.  I wonder if, as the sticks are returned to centre, and cross the centre, accumulated iTerm remains present, and won't go away until the sticks move to the opposite side.  It then 'comes back', which pushes the quad the opposite way from what would be expected from the user input.  I wonder if it may be better to zero out iTerm when returning to centre, so that iTerm won't be in the wrong direction when the sticks cross the middle.  This introduces its own issues, though, and adds complexity.  Hence I'm not really sure how to solve this specific issue.  It's only obvious when making small stick inputs, perhaps for 'perching', or to stabilise a quad when the battery is below.  Maybe having an option to retain iTerm may help, but then iTerm accumulation could cause massive motor boost because there was no motor_output limit.

Very tricky.

Do we have any flight videos or user testing at this point to confirm that it really works well?

ctzsnooze commented on src/main/blackbox/blackbox.c:
```diff
@@ -1405,6 +1405,8 @@ static bool blackboxWriteSysinfo(void)
         BLACKBOX_PRINT_HEADER_LINE(PARAM_NAME_TPA_BREAKPOINT_LOWER_FADE, "%d", currentPidProfile->tpa_breakpoint_lower_fade);
+        BLACKBOX_PRINT_HEADER_LINE(PARAM_NAME_EZ_LANDING_THRESHOLD, "%d",   currentPidProfile->ez_landing_threshold);
+        BLACKBOX_PRINT_HEADER_LINE(PARAM_NAME_EZ_LANDING_LIMIT, "%d",       currentPidProfile->ez_landing_limit);
         BLACKBOX_PRINT_HEADER_LINE("rc_rates", "%d,%d,%d",                  currentControlRateProfile->rcRates[ROLL],
```
Maybe it might help to include `mixer_type` in the log header, so we know if it is enabled or not?

tbolin replied:
mixer type added to log header

tbolin commented:
> I noticed that iTerm is zero when the pitch/roll stick is centered, and throttle is low. This will definitely help prevent bouncy landings where iTerm can be a contributor, and help 'un-sticking' a whoop from the wall... but it will make 'perching' very difficult, since adequate iTerm at low throttle is essential to hold the quad at the requested angle.
I also found something very strange that I couldn't really explain. Normally, under all zero throttle conditions, when I roll right, motors 3 and 4 increase. When I then hold a while, and sweep left across centre, the driven motors shift across until 1 and 2 are higher, when the roll stick is to the left of centre.
But if this PR's version of EZLanding is flashed, something very strange happens once airmode is active. If I activate airmode, put throttle to zero, and move the stick out to the right side, motors 3 and 4 spin up like normal. However, if I then try to centre the stick, motors 3 and 4 get stronger and stronger !! Very much NOT what I'd want. Then only once I get the stick back across centre do they finally slow down, and go the other way.
This makes for completely the opposite effect from what is intended, I would have thought. I don't know if this is a practical issue, but if I was correcting my position with sticks to the right, I would not want the motors to go harder when I was going back to centre, but that's what happens.
Here are two videos, zipped, trying to explain what I found. Without seeing the stick inputs it may not be sensible, but is easily replicated on any quad with this PR and props off.

What you are seeing is not the iTerm or iTerm gain being zeroed. It's the iTerm anti-windup kicking in since the output is saturated. That is the difference between the highest and lowest motor  is as high as it can get without raising the average to above the allowed limit.

In the video, when you bring the stick to the right, the iTerm start to accumulate. When the stick is then brought back to the centre, the max average output is again limited. The iTerm is now large enough to saturate the output on it's own, so the output is scaled down so that the average motor output stays at the limit (which is now at *ez_landing_limit* + whatever dynamic idle adds).
When the stick is moved to the left the limit is increased due to the stick being away from the centre. The iTerm left from when the stick was to the right is still there so the newly allowed authority is initially used to throttle up the "wrong" motors, until the iTerm has gone down enough for the PID output to change sign.

Also, if the iTerm anti-windup was not kicking in there still wouldn't be a fly away. The average motor output would still be limited (and therefore the maximum motor output would also be implicitly limited). The iTerm would however start accumulating, but the outputs wouldn't be changed until throttling up or moving the sticks away from the middle.
In an earlier version where the anti-windup did not kick in when it was supposed to the bounce prevention was still effective as can be seen in these logs from Sek101 back in january (where the bug was discovered. I now realized I should have posted more back then). 
[EzLanding Sek101 testing windup.zip](https://github.com/betaflight/betaflight/files/13610670/EzLanding.Sek101.testing.windup.zip)

I'll do some more bench testing now and see what I find.

tbolin commented:
also, the version where you can switch between the throttle based and clipping based mode can still be flashed by entering the entire commit hash of the commit before the clipping based mode was removed.

tbolin commented:
I did some bench testing with the same setup: First pushing the throttle up for a moment to activate airmode, and then slightly wiggling the roll stick back and forth. I did this with both the clipping mode and throttle mode.

Here are the logs:  
[EzLanding_bench_windup.zip](https://github.com/betaflight/betaflight/files/13611033/EzLanding_bench_windup.zip)

To me the results are very similar. The main differences I see are:
1. The iTerm stops accumulating when the outputs are saturated in the throttle based mode. The iTerm still reaches 40% at the end of the test in both modes.
2. The average output is slightly higher in the throttle based mode with sticks centred. This is expected since the throttle based mode accounts for dynamic idle.
3. When the sticks are centred, the differences between motors are retained in the throttle based mode, while all motors except one are at the same output in the clipping based mode. This is almost the definition of the difference between the modes, but I thought it was worth mentioning since I noticed the throttle based mode "sounded" worse. That is probably due to more frequencies being involved.
4. When the stick is pushed away from the centre two of the motors are still at the exact same output due to the motor output being clipped.

tbolin commented:
> I'm not sure why this PR gets that strange behaviour once airmode is active with small stick inputs. I wonder if, as the sticks are returned to centre, and cross the centre, accumulated iTerm remains present, and won't go away until the sticks move to the opposite side. It then 'comes back', which pushes the quad the opposite way from what would be expected from the user input. I wonder if it may be better to zero out iTerm when returning to centre, so that iTerm won't be in the wrong direction when the sticks cross the middle. This introduces its own issues, though, and adds complexity. Hence I'm not really sure how to solve this specific issue. It's only obvious when making small stick inputs, perhaps for 'perching', or to stabilise a quad when the battery is below. Maybe having an option to retain iTerm may help, but then iTerm accumulation could cause massive motor boost because there was no motor_output limit.

I think the best way to fix this would be to make the anti-windup non-sticky (I remember discussing this on discord over a year ago). Currently the iTerm is prevented from changing at all when the motor output range is saturated. If the iTerm was instead allowed to decrease as normal, but not increase during saturation this should be less of a problem. 
With the throttle based mode the iTerm can get "pinched" when the sticks go back to center. That is, the iTerm it self is enough to cause saturation, so it gets stuck at that value until the allowed authority is increased again (or the PID error gets so large in the other direction that the total output is no longer saturating).
With the "Legacy" mixer pinching can't happen since the default maximum for the iTerm (40%) isn't enough to cause saturation on it's own.

If the iTerm is allowed to decrease during saturation it would slowly go back to not saturating the output due to noise with throttle at min and sticks centered. It should also cause the weird behavior were the output first goes in the wrong direction to decrease, since the stick input in the opposite direction would be allowed to start decreasing the iTerm slightly earlier.

I can't think of a scenario where letting the iTerm decrease during saturation would be negative. From what I have seen in logs quads spends very little time at saturation under normal circumstances anyway and the instances where there is saturation and the sign of the iTerm is different from that of the error should be very rare. However, I remember that you had some different opinions on this.

Another way would be to scale the iTerm maximum by the same factor as the max motor output difference. A bit cruder, but would get the job done. Lowering the maximum could however erase some "legit" iTerm, but much better than resetting the iTerm (which I think would be worse than leaving things as they are). 
I was looking in to doing this but found that the implementation would be pretty crude so I left it out unless I could find some instance where it was an actual problem, which I never did.

haslinghuis commented:
@tbolin please rebase and resolve conflicts.

sugaarK reviewed (approved):
Approving untested

ctzsnooze commented:
@tbolin thanks for all the detailed analysis above.
We can look at suppressing iTerm differently later, it's a complicated business.  I agree that it may have a place in anti-bounce situations.  
Last night I flew this PR on a real quad, and, in practice, couldn't cause anything bad to happen.
The 'on-the-desk' strangeness didn't translate to anything bad now matter how carefully I tried to replicate the circumstances.
Most bumpy landings were a lot better controlled than normal.  It worked very well.
Perching was not so bad because P and D kept the attitude stable, the iTerm suppression wasn't much of a problem
The code is quite throttle sensitive, so retaining just a tiny bit of throttle made a perch quite stable.
I didn't exhaustively test the outcome with zero throttle flat drops, or inverted zero throttle drops.  I did some LOS, and they were OK. I'd expect to see them be a bit 'loose', FPV, unless a little throttle is retained, but that's to be tested later.
LOS flips and rolls appeared pretty much normal.
All up it works well - congratulations!
The code is well-encapsulated and won't cause issues if not enabled.
Hence I recommended merging, despite the late hour, and it now has been merged into 4.5!
Congratulations, and my apologies for being tardy in doing the testing.

ctzsnooze commented:
@tbolin please review the information provided on the release notes, to check I got it right>
https://github.com/betaflight/betaflight.com/blob/e9fb6859e94cdfada3707e353e06e34a9821e861/docs/release/Betaflight-4.5-Release-Notes.md

tbolin commented:
@ctzsnooze look good.

themeadery commented:
Tested this out today on 4.5.0-RC2 and I just wanted to say it works awesome! I was dying for this feature for so long and kept following all the progress. This implementation seems to be great. I tested it with defaults on the limit and threshold. I even gave it a torture test by landing on a bouncy bed mattress. Only one time it went into a bouncy oscillation, but I really had to plop it down wrong to make it do that. So awesome to be able to land smoothly like 6 times in a row if I want in one flight instead of disarming an inch above ground and slamming my battery/frame.
